src/toil_lib/tools/preprocessing.py

```diff
@@ -82,221 +84,425 @@ def run_samtools_index(job, bam_id):
+
+    # Call: picardtools
+    command = ['MarkDuplicates',
+               'INPUT=sorted.bam',
+               'OUTPUT=mkdups.bam',
+               'METRICS_FILE=metrics.txt',
+               'ASSUME_SORTED=true',
+               'CREATE_INDEX=true',
+               'VALIDATION_STRINGENCY=LENIENT']  # Ignores minor formatting issues
+
+    docker_call(work_dir=work_dir,
+                parameters=command,
+                # picard-tools container doesn't have JAVA_OPTS variable
+                # Set TMPDIR to /data to prevent writing temporary files to /tmp
+                env={'_JAVA_OPTIONS': '-Djava.io.tmpdir=/data/ -Xmx{}'.format(job.memory)},
+                tool='quay.io/ucsc_cgl/picardtools:1.95--dd5ac549b95eb3e5d166a5e310417ef13651994e',
+                outputs={'mkdups.bam': None, 'mkdups.bai': None})
+
+    bam = job.fileStore.writeGlobalFile(os.path.join(work_dir, 'mkdups.bam'))
+    bai = job.fileStore.writeGlobalFile(os.path.join(work_dir, 'mkdups.bai'))
+    return bam, bai
+
+
+def run_gatk_preprocessing(job, bam, bai, ref, ref_dict, fai, g1k, mills, dbsnp, unsafe=False):
+    """
+    GATK Preprocessing Pipeline
+    0: Mark Duplicates
+    1: Create INDEL realignment intervals
+    2: Realign INDELS
+    3: Recalibrate Base Quality Scores
+    4: Recalibrate Reads
+
+    :param JobFunctionWrappingJob job: passed automatically by Toil
+    :param str bam: FileStoreID for BAM file
+    :param str bai: FileStoreID for BAM index file
+    :param str ref: FileStoreID for reference genome fasta file
+    :param str ref_dict: FileStoreID for reference sequence dictionary file
+    :param str fai: FileStoreID for reference fasta index file
+    :param str g1k: FileStoreID for 1000 Genomes VCF file
+    :param str mills: FileStoreID for Mills VCF file
+    :param str dbsnp: FileStoreID for dbSNP VCF file
+    :param bool unsafe: If True, runs GATK tools in UNSAFE mode: "-U ALLOW_SEQ_DICT_INCOMPATIBILITY"
+    :return: FileStoreIDs for BAM and BAI files
     :rtype: tuple(str, str)
     """
-    rtc = job.wrapJobFn(run_realigner_target_creator, bam, bai, ref, ref_dict,
-                        fai, phase, mills, mem, unsafe, cores=job.cores)
-    ir = job.wrapJobFn(run_indel_realignment, rtc.rv(), bam, bai, ref, ref_dict,
-                       fai, phase, mills, mem, unsafe, cores=job.cores)
-    br = job.wrapJobFn(run_base_recalibration, ir.rv(0), ir.rv(1), ref, ref_dict,
-                       fai, dbsnp, mem, unsafe, cores=job.cores)
-    pr = job.wrapJobFn(run_print_reads, br.rv(), ir.rv(0), ir.rv(1), ref, ref_dict,
-                       fai, mem, unsafe, cores=job.cores)
-    # Wiring
-    job.addChild(rtc)
-    rtc.addChild(ir)
-    ir.addChild(br)
-    br.addChild(pr)
-    return pr.rv(0), pr.rv(1)
-
-
-def run_realigner_target_creator(job, bam, bai, ref, ref_dict, fai, phase, mills, mem, unsafe=False):
+    # MarkDuplicates runs best when Xmx <= 10G
+    mdups_memory = min(10737418240, job.memory)
```

Comment:
I wouldn't have this. Instead, I would throw a warning if the job has less than 10G of memory. Many batch systems will auto-kill a job whose resource allocations exceed their resource requests.

(Aside: is job.memory guaranteed to be an Int? Can't it be something like "10G"?)